var/spack/repos/builtin/packages/kokkos/package.py

```diff
@@ -2,231 +2,276 @@
-            # Host architectures
-            host_arch = spec.variants['host_arch'].value
-            # GPU architectures
-            gpu_arch  = spec.variants['gpu_arch'].value
-            if host_arch != 'none':
-                arch_args.append(host_arch)
-            if gpu_arch != 'none':
-                arch_args.append(gpu_arch)
-            # Combined architecture flags
-            if arch_args:
-                g_args.append('--arch={0}'.format(','.join(arch_args)))
-
-            # CUDA options
-            if '+force_uvm' in spec:
-                cuda_options_args.append('force_uvm')
-            if '+use_ldg' in spec:
-                cuda_options_args.append('use_ldg')
-            if '+rdc' in spec:
-                cuda_options_args.append('rdc')
-            if '+enable_lambda' in spec:
-                cuda_options_args.append('enable_lambda')
-            if cuda_options_args:
-                g_args.append('--with-cuda-options={0}'
-                              .format(','.join(cuda_options_args)))
-
-            # Kokkos options
-            if '+aggressive_vectorization' in spec:
-                kokkos_options_args.append('aggressive_vectorization')
-            if '+disable_profiling' in spec:
-                kokkos_options_args.append('disable_profiling')
-            if '+disable_dualview_modify_check' in spec:
-                kokkos_options_args.append('disable_dualview_modify_check')
-            if '+enable_profile_load_print' in spec:
-                kokkos_options_args.append('enable_profile_load_print')
-            if '+compiler_warnings' in spec:
-                kokkos_options_args.append('compiler_warnings')
-            if '+disable_deprecated_code' in spec:
-                kokkos_options_args.append('disable_deprecated_code')
-            if '+enable_eti' in spec:
-                kokkos_options_args.append('enable_eti')
-            if kokkos_options_args:
-                g_args.append('--with-options={0}'
-                              .format(','.join(kokkos_options_args)))
-
-            generate(*g_args)
-            make()
-            make('install')
+    version('master',  branch='master')
+    version('3.0', url="https://github.com/kokkos/kokkos/archive/3.0.00.tar.gz",
+            sha256="c00613d0194a4fbd0726719bbed8b0404ed06275f310189b3493f5739042a92b")
+    version('3.1', url="https://github.com/kokkos/kokkos/archive/3.1.00.tar.gz",
+            sha256="b935c9b780e7330bcb80809992caa2b66fd387e3a1c261c955d622dae857d878",
+            default=True)
+
+    depends_on("cmake@3.10:", type='build')
+
+    devices_variants = {
+        'cuda': [False, 'Whether to build CUDA backend'],
+        'openmp': [False, 'Whether to build OpenMP backend'],
+        'pthread': [False, 'Whether to build Pthread backend'],
+        'rocm': [False, 'Whether to build AMD ROCm backend'],
+        'serial': [True,  'Whether to build serial backend'],
```

Comment:
What about `hip`?

Reply:
Yeah, I should add that and remove ROCm. I need to figure out how to make variants version-dependent.

Reply:
As a workaround, you can just conflict.

Reply:
hip is added, along with missing vega architectures. I added the conflict for versions where appropriate.